architecture-building-systems/CityEnergyAnalyst
# Use shapes tagged as `building:part` in OSM as well as `building` in `zone_helper`

This PR implements the fix suggested by @lguilhermers in #3220, whereby shapes tagged as `building:part` in OSM are imported when running the `zone_helper`. This is implemented as follows:
1. `building` shapes are imported as done in the zone_helper so far
2. `building:part` shapes are imported and merged to the same GeoDataFrame as the buildings
3. The combined shapefile is passed to the `fix_overlapping_geometries` script by @MatNif to ensure buildings split as required

I'm adding a Boolean parameter so users can decide whether they want to use building parts or not. Adding building parts means creating a lot of overlapping geometries, so I am making `fix_overlapping_geometries == True` if `include_building_parts == True`.

Here's the results of this PR:
- 'Shapefile created by the master branch with `fix_overlapping_geometries == False`:'
<img width="1280" alt="Screen Shot 2022-08-26 at 00 11 29" src="https://user-images.githubusercontent.com/18548065/186719103-f5dbf192-daab-457a-9760-8c14f08a7524.png">
- 'Shapefile created after this PR:'
<img width="1280" alt="Screen Shot 2022-08-26 at 00 07 13" src="https://user-images.githubusercontent.com/18548065/186719210-437237fa-e6be-4df7-827a-3024b5dafff7.png">
Note:

1. Lau Pa Sat was originally imported as a  single polygon with no height information – now it is imported as a number of overlapping polygons (see #3236)
2. Two large towers were being imported as much shorter before (solved by #3222), and the complex shape of the Asia Square towers were imported as two blocks with the wrong height
3. Robinson Centre was originally imported as a single big block, even though it's actually a podium and a tower – the new version still has the same volume, but the podium is a separate polygon for which users can adjust the height as required.

In solving this issue, I ended up discovering a few others, as discussed in #3220 and #3232, including the fact that by using `building:part`s some information from OSM was being lost. So now I have added a step to copy any building use type available in the OSM `building` types to the OSM `building:part` types (which generally have no use type information). I furthermore noticed some inconsistencies in the way typologies were assigned:
1. The CEA code claims to "Calculate the most likely use type" in `calculate_typology_file`, but actually any data not imported from OSM is just assigned "MULTI_RES". So I added a step to assign the most common use type to the entire case study area to those buildings (similar to the median used in the heights and number of floors, but in this case I had to use to mode)
2. I also noticed that the `amenity` information from OSM was not being used in assigning the use types, which means that , for example, food courts were being assigned as "RETAIL" (rather than, say, "RESTAURANT"). I have rectified this as well.

How to test this PR:
1. Select an area where OSM includes both polygons tagged as buildings and polygons tagged as building:part – ideally not any of the examples shown above _and_ where the polygon's geometry is not `None` (which was the case or the Zurich example I mentioned in #3220), though finding such an area might be tricky
2. Run the zone_helper
4. Check whether building and building:part polygons are imported as separate buildings

I'm assigning @lguilhermers as a reviewer since he originally requested this feature and had many ideas on how to implement it. I'm also assigning @MatNif since getting this PR to work required some reworking of his code to fix overlapping geometries.

This PR requires #3232 to be merged first and closes #3220
